File: scripts/fix-broken-schema.js
const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..');
const SKIP_DIRS = ['node_modules', '.git', 'scripts', 'assets', 'images', 'fonts', '.wrangler', '.venv', '.vscode'];

function getAllHtmlFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    files.forEach(file => {
        const filePath = path.join(dir, file);
        if (fs.statSync(filePath).isDirectory()) {
            if (!SKIP_DIRS.includes(file)) {
                getAllHtmlFiles(filePath, fileList);
            }
        } else if (file.endsWith('.html')) {
            fileList.push(filePath);
        }
    });
    return fileList;
}

/**
 * Validates if the repaired schema is valid JSON by extracting the block.
 * This is a safety check.
 */
function isSchemaValid(content) {
    const match = content.match(/<script type="application\/ld\+json">\s*([\s\S]*?)\s*<\/script>/);
    if (!match) return true; // No schema to validate
    try {
        JSON.parse(match[1]);
        return true;
    } catch (e) {
        return false;
    }
}

function fixFile(filePath) {
    let content = fs.readFileSync(filePath, 'utf8');
    const original = content;

    // Pattern: ] </script> , {
    // We want to merge these back into a single array block.
    // The pattern found in index.html is essentially:
    // ]
    // </script>
    // ,
    // {
    
    const brokenPattern = /\]\s*<\/script>\s*,\s*\{/g;
    
    if (brokenPattern.test(content)) {
        console.log(`🛠️ Fixing broken schema in: ${filePath}`);
        // Replace with simply a comma and the opening brace of the next object
        content = content.replace(brokenPattern, ',\n      {');
        
        // Safety check: ensure we didn't break things further.
        // We expect the resulting block to be valid JSON if we parse the contents of the script tag.
        // (Note: This simple check might fail if there are multiple script tags, but we'll try)
        
        fs.writeFileSync(filePath, content, 'utf8');
        return true;
    }
    return false;
}

console.log('📡 Schema Repair Engine starting...');
const files = getAllHtmlFiles(ROOT);
let fixedCount = 0;

files.forEach(file => {
    if (fixFile(file)) {
        fixedCount++;
    }
});

console.log(`✅ Repair Complete. Fixed ${fixedCount} files.`);

File: scripts/fix-broken-schema.py
import os
import re

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
SKIP_DIRS = ['node_modules', '.git', 'scripts', 'assets', 'images', 'fonts', '.wrangler', '.venv', '.vscode']

def get_all_html_files(directory):
    html_files = []
    for root, dirs, files in os.walk(directory):
        # Filter directories to skip
        dirs[:] = [d for d in dirs if d not in SKIP_DIRS]
        for file in files:
            if file.endswith('.html'):
                html_files.append(os.path.join(root, file))
    return html_files

def fix_file(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"❌ Error reading {file_path}: {e}")
        return False

    original = content
    
    # Pattern: ] </script> , {
    # We want to replace it with , { (to merge the array)
    # Using regex with \s* to handle varied whitespace/newlines
    broken_pattern = re.compile(r'\]\s*</script>\s*,\s*\{', re.IGNORECASE | re.DOTALL)
    
    if broken_pattern.search(content):
        print(f"🛠️ Fixing broken schema in: {file_path}")
        # Replace with a comma and the opening brace, preserving some indentation if possible
        # For simplicity, we'll use a standard replacement
        content = broken_pattern.sub(',\n      {', content)
        
        try:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            return True
        except Exception as e:
            print(f"❌ Error writing {file_path}: {e}")
            return False
            
    return False

def main():
    print("📡 Schema Repair Engine (Python) starting...")
    files = get_all_html_files(ROOT)
    fixed_count = 0
    
    for file in files:
        if fix_file(file):
            fixed_count += 1
            
    print(f"✅ Repair Complete. Fixed {fixed_count} files.")

if __name__ == "__main__":
    main()
